iOS: Provide correct initial value for AppState.currentState

rigdern opened the pull request:
Attempt to fix #7919.

Currently, if the app is launched into the background and you read `AppState.currentState` too soon, you will see the value `'active'` instead of `'background'`. This is because the default value of `AppState.currentState` is hardcoded to be `'active'` and it is initialized with the actual value asynchronously.

This attempts to fix the bug by having the `RCTAppState` module provide the initial state as a module constant.

As noted in #7919, it looks like this fix was already tried and reverted with 0fb3d8de83c4d8b79a8a110e9cd1300b39ab8918. @zjj010104, @hedgerwang, @nicklockwood -- can you explain why? I would very much like to get this bug fixed. Nobody has followed up on the issue I filed so I decided to just go ahead and make a PR with my best guess at a fix.

**Test plan (required)**

Built a small app as described in the repro steps for #7919 and verified that, when the app is launched into the background, `init currentState: background` is printed. Also verified that  `init currentState: active` is printed when the app is launched into the foreground.

Adam Comella
Microsoft Corp.

facebook-github-bot-4 commented:
By analyzing the blame information on this pull request, we identified @nicklockwood and @skv-headless to be potential reviewers.

janicduplessis commented on Libraries/AppState/AppState.js:
```diff
@@ -86,9 +86,10 @@ class AppState extends NativeEventEmitter {
     // TODO: getCurrentAppState callback seems to be called at a really late stage
```
Maybe remove this comment now.

rigdern replied:
removed

janicduplessis commented:
This looks good to me, we might want to check with @hedgerwang @zjj010104 why something similar was reverted before shipping it. Although this implementation seems safer as it will use 'active' if the initial constant is null for some reason.

mkonicek commented:
@javache would you mind taking a look please? Just trying to triage impactful PRs, splitting iOS ones between you and Martin.

javache commented:
Just an update: I'm figuring out internally what the issue where that caused the previous revert. Will keep you updated.

javache commented:
One potential issue with this change. At FB we use inline/lazy requires which means this module will only be instantiated when its first used. `constantsToExport` however is invoked at bridge startup. If the app launches in the background and you foreground, there won't actually be anyone listening to this update and the state cached in this module will be incorrect.

rigdern commented:
@javache thanks for looking into the issue. Do you have any recommendations for a fix?

The TODO comments in [AppState.js](https://github.com/rigdern/react-native/blob/ef995a2fbf22fa143f71e8c58dee59c887910525/Libraries/AppState/AppState.js) mention deprecating the synchronous property and only supporting an asynchronous method to get the current state.

However, my app needs to have this value synchronously during app launch. If the app is launching into the background, we don't want to render any UI. If the app is launching into the foreground, we need to smoothly transition between the launch image and the initial app UI (i.e. we don't want the transition to result in a flicker). In the launch into foreground case, we need to have the app state synchronously otherwise there will be a flicker due to the delay before starting to render the UI.

My current PR doesn't sound good because, as you pointed out, the app state can be stale if the AppState module was loaded lazily.

If we have to live with the constraint that the AppState module can be loaded lazily, I don't see how we can have an accurate synchronous API for retrieving the current app state. Would it be possible and desirable to always have the AppState module loaded eagerly? Then, after my PR, `currentState` will always be accurate and available synchronously.

Thoughts?

rigdern commented:
ping @javache

rigdern commented:
ping @javache

facebook-github-bot-7 commented:
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @nicklockwood as a potential reviewer. Could you take a look please or cc someone with more context?

ide commented on Libraries/AppState/AppState.js:
```diff
@@ -86,9 +86,10 @@ class AppState extends NativeEventEmitter {
-    
+    // If RCTAppState implements initialAppStateInitialize, use that. Otherwise,
+    // use 'active'.
+    this.currentState = RCTAppState.initialAppState || 'active';
```
Should be able to remove the "active" fallback right?

rigdern replied:
I think we need to keep it until we export a `initialAppState` constant on Android.

ide replied:
Ah good point, can you explicitly call that out in a comment?

rigdern replied:
done

ide commented:
This seems fine to me in that due to the asynchronous nature of the RN bridge, there is always a chance that the JS's AppState will be stale (not yet up-to-date). The important thing I think is achieving eventual consistency. 

That is, if you subscribe to AppState events (from your app's own JS), even if `AppState. currentState` is stale at the moment (which is rare in practice I'd imagine -- most of the time your app's scaffolding is subscribing to AppState), you should quickly receive an event with the correct state.

rigdern commented:
@ide Thanks for taking a look. I addressed all of the feedback.

I think you are right that it's not a big deal if `currentState` is stale. At least now whatever value `currentState` has, it represents either the current state or a previous state of the application. Whereas today, if you read the hard coded fallback value it can represent a state the application was never actually in.

ide commented:
This seems correct - as soon as AppState is required, it will request the latest value from the native side and subscribe to new changes, and in the meantime we'll use the potential stale (but once-correct) currentState.

@facebook-github-bot shipit

facebook-github-bot-9 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/282518032094909/int_phab) to review.

rigdern commented:
@ide, thanks for reviewing this and getting it merged!
